client/x/evmengine/types/tx_test.go

```diff
@@ -162,3 +169,270 @@ func TestEVMEvent_Verify(t *testing.T) {
 		})
 	}
 }
+
+func TestEthLogToEVMEvent(t *testing.T) {
+	t.Parallel()
+	upgradeAbi := initializeABI(t)
+	data, err := upgradeAbi.Events["SoftwareUpgrade"].Inputs.NonIndexed().Pack("test-upgrade", int64(1), "test-info")
+	require.NoError(t, err)
+
+	tcs := []struct {
+		name           string
+		log            ethtypes.Log
+		expectedResult *types.EVMEvent
+		expectedErr    string
+	}{
+		{
+			name: "pass: zero address",
+			log: ethtypes.Log{
+				Address: emptyAddr,
+				Topics:  []common.Hash{types.SoftwareUpgradeEvent.ID},
+				Data:    data,
+			},
+			expectedResult: &types.EVMEvent{
+				Address: emptyAddr.Bytes(),
+				Topics:  [][]byte{types.SoftwareUpgradeEvent.ID.Bytes()},
+				Data:    data,
+			},
+		},
+		{
+			name: "pass: empty data",
+			log: ethtypes.Log{
+				Address: dummyContractAddress,
+				Topics:  []common.Hash{types.SoftwareUpgradeEvent.ID},
+				Data:    emptyData,
+			},
+			expectedResult: &types.EVMEvent{
+				Address: dummyContractAddress.Bytes(),
+				Topics:  [][]byte{types.SoftwareUpgradeEvent.ID.Bytes()},
+				Data:    emptyData,
+			},
+		},
+		{
+			name: "pass: zero address & empty data",
+			log: ethtypes.Log{
+				Address: emptyAddr,
+				Topics:  []common.Hash{types.SoftwareUpgradeEvent.ID},
+				Data:    emptyData,
+			},
+			expectedResult: &types.EVMEvent{
+				Address: emptyAddr.Bytes(),
+				Topics:  [][]byte{types.SoftwareUpgradeEvent.ID.Bytes()},
+				Data:    emptyData,
+			},
+		},
+		{
+			name: "pass: full log",
+			log: ethtypes.Log{
+				Address: dummyContractAddress,
+				Topics:  []common.Hash{types.SoftwareUpgradeEvent.ID},
+				Data:    data,
+			},
+			expectedResult: &types.EVMEvent{
+				Address: dummyContractAddress.Bytes(),
+				Topics:  [][]byte{types.SoftwareUpgradeEvent.ID.Bytes()},
+				Data:    data,
+			},
+		},
+		{
+			name: "fail: empty topics",
+			log: ethtypes.Log{
+				Address: dummyContractAddress,
+				Topics:  []common.Hash{},
+				Data:    data,
+			},
+			expectedErr: "verify log: empty topics",
+		},
+	}
+
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			result, err := types.EthLogToEVMEvent(tc.log)
+			if tc.expectedErr != "" {
+				require.EqualError(t, err, tc.expectedErr)
+			} else {
+				require.NoError(t, err)
+				require.Equal(t, tc.expectedResult, result)
+			}
+		})
+	}
+}
+
+func TestSortEVMEvents(t *testing.T) {
+	t.Parallel()
+	stakingAbi, err := bindings.IPTokenStakingMetaData.GetAbi()
+	require.NoError(t, err, "failed to load ABI")
+	stakingAddr := common.HexToAddress(predeploys.IPTokenStaking)
+	slashingAbi, err := bindings.IPTokenSlashingMetaData.GetAbi()
+	require.NoError(t, err, "failed to load ABI")
+	slashingAddr := common.HexToAddress(predeploys.IPTokenSlashing)
+	require.Negative(t, bytes.Compare(stakingAddr.Bytes(), slashingAddr.Bytes()), "stakingAddr should be less than slashingAddr")
```

Comment:
What do you mean `stakingAddr should be less than slashingAddr` for a compare? (+ same for below error messages)

Reply:
Thanks for the feedback! The reason for checking that `stakingAddr` is less than `slashingAddr` is to ensure a known initial order before testing the sorting logic. 
This precondition is necessary because we want to verify that if the addresses are input in the wrong order, the sorting function will correctly reorder them. 
Similarly, the same check is applied for the event topics (withdrawEv and depositEv) to confirm the function’s behavior.

To make this clearer, I’ve added more detailed comments in the code explaining why these comparisons are necessary and included examples to illustrate the expected behavior. (commit: https://github.com/piplabs/story/pull/106/commits/5027d58e7053f92f5ae329cd74dc8a73e575e295) 
Please take a look at the updated comments, and let me know if there’s anything else I can clarify.